lib/streamlit/elements/json.py

```diff
@@ -70,6 +71,18 @@ def json(
 
         if not isinstance(body, str):
             try:
+                # Convert sets into lists, which render more nicely on the frontend
+                set_found = False
+                for key in body:
+                    if isinstance(body[key], set):
+                        if not set_found:
+                            # When set is found to prevent mutation of input body, we need to deepcopy it once
+                            # To avoid copying it multiple times we use set_found flag
+                            body = copy.deepcopy(body)
+                            set_found = True
+                        body[key] = list(body[key])
+
```

Comment:
Let's do the sets-to-lists conversion outside of this try-except block. (The `try` is there to catch `json.dumps` errors specifically - if there is an issue with the sets-to-lists conversion, we shouldn't be swallowing the error here.)

Reply:
Also, this is definitely a minor nit so take it more as a suggestion, but: it might be nice to factor this logic out into its own function to make things easier to read. For example:

```python
def _convert_sets_to_lists(body: Any) -> Any:
    # Convert sets into lists, which render more nicely on the frontend
    set_found = False
    for key in body:
        if isinstance(body[key], set):
            if not set_found:
                # When set is found to prevent mutation of input body, we need to deepcopy it once
                # To avoid copying it multiple times we use set_found flag
                body = copy.copy(body)
                set_found = True
            body[key] = list(body[key])

class JsonMixin:
    def json(...):
       # other stuff
        if not isinstance(body, str):
            body = _convert_sets_to_lists(body)
            try:
                # Serialize body to string
                body = json.dumps(body, default=repr)
            except TypeError as err:
                st.warning(
                    "Warning: this data structure was not fully serializable as "
                    "JSON due to one or more unexpected keys.  (Error was: %s)" % err
                )
                body = json.dumps(body, skipkeys=True, default=repr)
        # other stuff
```

Reply:
Thank you for your review 👍  I've done both:

- code is move out of `try` block
- code is inside `function`